src/utils/AuthProvider.tsx

```diff
@@ -0,0 +1,116 @@
+'use client';
+
+import { AuthError, AuthResponse, Session } from '@supabase/supabase-js';
+import { createContext, useContext, useEffect, useMemo, useState } from 'react';
+import { UUID } from 'crypto';
+import supabase from '@/api/supabase/createClient';
+
+// TODO: resolve inconsistent null/undefined types for empty values
+export interface AuthContextType {
+  session: Session | undefined;
+  userId: UUID | undefined;
+  userEmail: string | undefined;
+  signIn: (email: string, password: string) => Promise<AuthResponse>;
+  signOut: () => Promise<AuthError | null>;
+  signUp: (
+    email: string,
+    password: string,
+    options: object,
+  ) => Promise<AuthResponse>;
+}
+
+export const AuthContext = createContext<AuthContextType | undefined>(
+  undefined,
+);
+
+// TODO: we might need error handling here?
+export const useAuth = () => useContext(AuthContext);
+
+export default function AuthProvider({
+  children,
+}: {
+  children: React.ReactNode;
+}) {
+  const [session, setSession] = useState<Session | undefined>(undefined);
+  const [userId, setUserId] = useState<UUID | undefined>(undefined);
+  const [userEmail, setUserEmail] = useState<string | undefined>(undefined);
+
+  // set the session, userId, and userEmail
+  const setAll = (newSession: Session | undefined) => {
+    if (!newSession) return;
+    setSession(newSession);
+    if (newSession?.user?.id) {
+      const sessionUserId = newSession.user.id as UUID;
+      const sessionUserEmail = newSession.user.email;
+      setUserId(sessionUserId);
+      setUserEmail(sessionUserEmail);
+    }
+  };
+
+  // on page load, check if there's a session and set it
+  useEffect(() => {
+    supabase.auth.getSession().then(({ data: { session: newSession } }) => {
+      setAll(newSession != null ? newSession : undefined);
+    });
+
+    supabase.auth.onAuthStateChange((_event, newSession) => {
+      setAll(newSession != null ? newSession : undefined);
+    });
+  }, []);
+
+  // sign in and set the session, userId, and userEmail
+  const signIn = async (email: string, password: string) => {
+    const value = await supabase.auth.signInWithPassword({
+      email,
+      password,
+    }); // will trigger onAuthStateChange to update the session
+
+    setAll(value.data.session ? value.data.session : undefined);
+
+    return value;
+  };
+
+  // sign out and clear the session, userId, and userEmail
+  const signOut = async () => {
+    const { error } = await supabase.auth.signOut();
+
+    setSession(undefined);
+    setUserId(undefined);
+    setUserEmail(undefined);
+
+    return error;
```

Comment:
hmmm yeah i have no idea -- seems like the only one on [here](https://supabase.com/docs/reference/javascript/auth-signout) is you can't sign out if you're not already signed in, which shouldn't be possible anyways because of the way this is designed (i.e. sign out button doesn't show up if you're not logged in)

i think gwn does not check for the sign out error? not sure if we can just skip it

but yes im changing it so it only sets things to undefined if there is no error

Reply:
ohh that makes sense

on a slightly less relevant tangent --
looking at the docs, i noticed this blurb:
> By default, `signOut()` uses the global scope, which signs out all other sessions that the user is logged into as well.

given this, should we add a check in `onAuthStateChange` to set those fields to be `undefined` when we receive a log out event? In fact, should we just move all our state changes into the `onAuthStateChange` function depending on the event received?

not sure if that would work, but it may make things a lot cleaner and secure

Reply:
@varortz thoughts on moving things to `onAuthStateChange`? im chill w anything though